File: sample-java/src/main/java/com/mapp/engagesample/inbox/InboxViewModel.java
package com.mapp.engagesample.inbox;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.appoxee.Appoxee;
import com.appoxee.shared.InboxMessage;
import com.appoxee.shared.MessageStatus;

import java.util.ArrayList;
import java.util.List;

public class InboxViewModel extends ViewModel {

    private final List<InboxMessage> internalMessages = new ArrayList<>();
    private final MutableLiveData<InboxStateUi> state = new MutableLiveData<>(new InboxStateUi(false));
    public LiveData<InboxStateUi> stateUi = state;

    public InboxViewModel() {
        state.setValue(new InboxStateUi(true));
        Appoxee.instance().fetchInboxMessages().enqueue(result -> {
            if (result.isSuccess() && result.getData() != null) {
                List<InboxMessage> messages = result.getData().getMessages();
                state.setValue(new InboxStateUi(updateMessages(messages)));
            } else {
                Throwable t = result.getError();
                String error = (t != null && t.getMessage() != null) ? t.getMessage() : "Unknown error";
                state.setValue(new InboxStateUi(error));
            }
        });
    }

    public void updateStatus(InboxMessage message, int index, MessageStatus messageStatus) {
        Appoxee.instance().updateInboxMessageStatus(message, messageStatus).enqueue(result -> {
            if (result.isSuccess()) {
                InboxMessage updatedMessage = message.setStatus(messageStatus);
                internalMessages.set(index, updatedMessage);
                InboxStateUi inboxState = new InboxStateUi(internalMessages);
                state.setValue(inboxState);
            }
        });
    }

    private List<InboxMessage> updateMessages(List<InboxMessage> messages) {
        internalMessages.clear();
        if (messages != null && !messages.isEmpty()) {
            internalMessages.addAll(messages);
        }
        return internalMessages;
    }
}

File: sample-kotlin/src/main/java/com/mapp/engagesample/inbox/InboxViewModel.kt
package com.mapp.engagesample.inbox

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.appoxee.Appoxee
import com.appoxee.internal.model.response.inbox.InboxMessage
import com.appoxee.internal.model.response.inbox.MessageStatus
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import okhttp3.internal.toImmutableList

class InboxViewModel : ViewModel() {
    private val tag = InboxViewModel::class.java.simpleName

    private val messages = mutableListOf<InboxMessage>()
    private val state = MutableStateFlow(InboxStateUI())
    val stateUi: StateFlow<InboxStateUI> = state
    private val coroutineContext =
        SupervisorJob() + Dispatchers.IO + CoroutineExceptionHandler { coroutineContext, throwable ->
            Log.e(tag, "EXCEPTION: $throwable")
            state.value = InboxStateUI(error = throwable.message)
        }

    init {
        viewModelScope.launch(coroutineContext) {
            state.value = InboxStateUI(isLoading = true)
            messages.clear()
            val result = Appoxee.instance().fetchInboxMessages().asSuspend()
            if (result.isSuccess()) {
                result.getData()?.messages?.let {
                    messages.addAll(it)
                }
                state.value = InboxStateUI(messages = messages.toImmutableList())
            } else {
                val error = result.getError()?.message
                state.value = InboxStateUI(error = error)
            }
        }
    }

    fun updateMessageStatus(message: InboxMessage, index:Int, messageStatus: MessageStatus) {
        state.value = InboxStateUI(isLoading = true)
        viewModelScope.launch(coroutineContext) {
            val result =
                Appoxee.instance().updateInboxMessageStatus(message, messageStatus).asSuspend()
            if (result.isSuccess()) {
                messages[index] = message.setStatus(messageStatus)
                state.value = InboxStateUI(messages = messages.toImmutableList())
            } else {
                val error = result.getError()?.message
                state.value = InboxStateUI(error = error)
            }
        }
    }
}